### DISCUSSION TITLE
Proposal to Add Completion Capability to MCP Specification

### DISCUSSION THREAD

**Discussion by gavinaboulhosn** [modelcontextprotocol/modelcontextprotocol#122] [Ideas - General]
### Pre-submission Checklist

- [X] I have verified this would not be more appropriate as a feature request in a specific repository
- [X] I have searched existing discussions to avoid duplicates

### Your Idea

I've observed that the current Model Context Protocol (MCP) specification does not provide a mechanism for servers to indicate whether they support completions. As a result, when a client requests completions from a server that doesn't implement this feature, the server responds with a "Method not found" error (`-32601`).

**Current Behavior:**

- MCP servers declare capabilities such as `resources`, `prompts`, and `tools` during initialization.
- There is no specific capability flag for `completions`.
- Attempting to request completions from unsupported servers results in a `-32601` error.

**Short-Term Solution:**

To address this issue, clients can implement error handling that track the `-32601` error. When this error is encountered, the client can disable further completion requests for the affected template or prompt, preventing repeated unsuccessful attempts.

**Long-Term Recommendation:**

For a more robust and scalable solution, I propose incorporating completions into the capability negotiation process. This enhancement would allow clients and servers to agree on supported features upfront, eliminating the need for error-based detection. 

**Proposed Capability Addition:**

```typescript
const server = new Server({
  name: 'My Server',
  version: '1.0.0',
}, {
  capabilities: {
    resources: { completions: true },
    prompts: { completions: true },
  }
});

// or

const server = new Server({
  name: 'My Server',
  version: '1.0.0',
}, {
  capabilities: {
       resources: {},
       prompts: {},
       experimental: {
        completions: { prompts: true, resources: true }
      }
  }
});
```

**Benefits:**

- **Improved Reliability:** Avoids reliance on error handling for feature detection, reducing potential bugs and enhancing user experience.
- **Enhanced Protocol Consistency:** Aligns the completion feature with existing capability-based feature negotiation, maintaining uniformity across different capabilities.
- **Better Developer Experience:** Clients can clearly understand server capabilities upfront, leading to more efficient and predictable integrations.

### Scope

- [X] Protocol Specification
- [ ] SDK Features
- [ ] Documentation
- [ ] Developer Experience
- [ ] Other

**Comment by ondrsh** [comment]
Even if this was implemented in the `Capabilities` layer, there is currently no way to signal which prompt arguments support completions and which don't. For example:

```json
{
  "name": "write_short_story",
  "description": "Produces a short story in a specified genre, in the style of a particular author.",
  "arguments": [
    {
      "name": "genre",
      "description": "The story's genre (e.g., 'fantasy', 'science-fiction', 'romance', etc.)",
      "required": false
    },
    {
      "name": "author",
      "description": "An author whose style you want to emulate (e.g., 'Tolkien', 'Lovecraft', or 'Jane Austen')",
      "required": false
    }
  ]
}
```

<br>

Let's say the server supports completions for `genre`, but not for `author`, because it doesn't have access to a database of author names. There is no way to know this from the client side. If `author` gets a completion request, do we return `-32602` or an empty list?

Maybe we could get away with always returning an empty list, but this would lead to lots of unnecessary requests. If we returned `-32602`, we could cache the error on the client side and not send requests for that argument again, but this feels like an ugly hack (thinking about all client implementations). An idea to signal this would be to add `supportsCompletion` to prompt arguments.

But this actually touches on a more general issue: The completion feature seems to be closely mirroring the LSP counterpart. The problem is that LSP servers are almost exclusively running locally. This won't be the case for MCP. The docs mention that implementations should handle rate limiting, but thinking about how this will play out in practice makes me wonder. Right now there is no MCP error code for this scenario, so the transport layers would need to bubble this up into the applications (which becomes messy).

**Comment by jspahrsummers** [comment]
Related: https://github.com/modelcontextprotocol/specification/issues/157

**Comment by localden** [comment]
We've limited Discussions in this repository to meeting notes and are closing this thread. To continue the conversation, pick a channel from [Contributor Communication](https://modelcontextprotocol.io/community/communication). Thanks for posting.
